### PR TITLE
Fix to honor service provider configurations

### LINKED ISSUE #9745: Kerberos grant always add the userstore and tenant domain to the subject identifier
**Describe the issue:**
The following configurations of the service provider are not honored in the Kerberos grant. It always adds the userstore and tenant domain to the subject identifier
- Use tenant domain in local subject identifier
- Use user store domain in local subject identifier

This can be fixed by setting the AuthenticatedSubjectIdentifier as below at [1]. setAuthenticatedSubjectIdentifier() will set the subject by considering the above two configurations.
```
kerberosUser.setAuthenticatedSubjectIdentifier(kerberosUser.getUserName(), serviceProvider);
```

[1] https://github.com/wso2-extensions/identity-oauth2-grant-kerberos/blob/master/src/main/java/org/wso2/carbon/identity/oauth2/grant/kerberos/KerberosGrant.java#L230-L232

### PR DESCRIPTION
**Pull request by LahiruLS**
Fix https://github.com/wso2/product-is/issues/9745

### CODE REVIEW

**Comment by isharak** [comment]
@LahiruLS can you please fix the conflicts 
@mefarazath can you please review this.

File: pom.xml [hunk 1]
```diff
@@ -45,7 +50,7 @@
         <dependency>
             <groupId>org.wso2.carbon</groupId>
             <artifactId>org.wso2.carbon.user.core</artifactId>
-            <version>4.4.3</version>
+            <version>4.5.1</version>
```
Comment (mefarazath) [inline pom.xml:53]: Bumping this dependency will make this extension not backward compatible. 
is this expected?

File: src/main/java/org/wso2/carbon/identity/oauth2/grant/kerberos/KerberosGrant.java [hunk 2]
```diff
@@ -383,4 +390,29 @@ private boolean isUserExistsInUserStore(String username, String tenantDomain, St
 
         return false;
     }
+
+    private ServiceProvider getServiceProvider(OAuth2AccessTokenReqDTO tokenReq) throws IdentityOAuth2Exception {
+        ServiceProvider serviceProvider;
+        try {
+            serviceProvider = OAuth2ServiceComponentHolder.getApplicationMgtService().getServiceProviderByClientId(
+                    tokenReq.getClientId(), OAuthConstants.Scope.OAUTH2, tokenReq.getTenantDomain());
```
Comment (mefarazath) [inline src/main/java/org/wso2/carbon/identity/oauth2/grant/kerberos/KerberosGrant.java:398]: OAuthConstants.Scope.OAUTH2 is not the constant to use here i guess. The value we need to pass is the inbound type we are looking up.

Use [1] or [2]

[1] https://github.com/wso2/carbon-identity-framework/blob/4e311da4e8e686bd74c8abcb17763f0ef61bd5b3/components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkConstants.java#L300

[2] https://github.com/wso2/carbon-identity-framework/blob/4e311da4e8e686bd74c8abcb17763f0ef61bd5b3/components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkConstants.java#L59

File: src/main/java/org/wso2/carbon/identity/oauth2/grant/kerberos/KerberosGrant.java [hunk 3]
```diff
@@ -383,4 +390,29 @@ private boolean isUserExistsInUserStore(String username, String tenantDomain, St
...
+        } catch (IdentityApplicationManagementException e) {
+            throw new IdentityOAuth2Exception("Error occurred while retrieving OAuth2 application data for client id " +
+                    tokenReq.getClientId(), e);
+        }
+        if (serviceProvider == null) {
+            if (log.isDebugEnabled()) {
+                log.debug("Could not find an application for client id: " + tokenReq.getClientId()
+                        + ", scope: " + OAuthConstants.Scope.OAUTH2 + ", tenant: " + tokenReq.getTenantDomain());
```
Comment (mefarazath) [inline src/main/java/org/wso2/carbon/identity/oauth2/grant/kerberos/KerberosGrant.java:406]:
```suggestion
                        + ", tenant: " + tokenReq.getTenantDomain());
```

File: src/main/java/org/wso2/carbon/identity/oauth2/grant/kerberos/KerberosGrant.java [hunk 4]
```diff
@@ -383,4 +390,29 @@ private boolean isUserExistsInUserStore(String username, String tenantDomain, St
...
+            }
+            throw new IdentityOAuth2Exception("Service Provider not found");
+        }
+        if (log.isDebugEnabled()) {
+            log.debug("Retrieved service provider: " + serviceProvider.getApplicationName() + " for client: " +
+                    tokenReq.getClientId() + ", scope: " + OAuthConstants.Scope.OAUTH2 + ", tenant: " +
```
Comment (mefarazath) [inline src/main/java/org/wso2/carbon/identity/oauth2/grant/kerberos/KerberosGrant.java:412]:
```suggestion
                    tokenReq.getClientId() + ", tenant: " +
```

**Review by mefarazath — COMMENTED** [review]
Added a few concern points.

**Comment by somindatommy** [comment]
@LahiruLS can you address the comments and fix the comments?

**Comment by LahiruLS** [comment]
I will address all of them in another Pull request. 

Thanks,
Lahiru

**Pull request closed without merging**